Fix misaligned pointer warnings

magnumripper opened the pull request:
Both had SALT_ALIGN fixed to 8 where it should be sizeof(pointer). I found no issues for them so fixed them rather than create issues.

magnumripper committed:
Fix a misaligned pointer warning in AndroidFDE

solardiz commented on src/androidfde_fmt_plug.c:
```diff
@@ -62,7 +62,7 @@ john_register_one(&fmt_fde);
 #define BINARY_ALIGN        1
-#define SALT_ALIGN          8
+#define SALT_ALIGN          sizeof(struct custom_salt*)
 #define SALT_SIZE           sizeof(struct custom_salt)
```
I think it's confusing to use `sizeof(struct custom_salt*)` here because that's not the pointer type found _in_ the struct. We actually need alignment because of the `unsigned char *cipherbuf;` field here (BTW, the struct could be made smaller by re-ordering fields, not keeping this pointer between `int` fields). So I think it's better to use `sizeof(void *)` or `sizeof(unsigned char *)`.

solardiz commented on src/opencl_pbkdf2_hmac_sha512_fmt_plug.c:
```diff
@@ -33,9 +33,9 @@ john_register_one(&fmt_opencl_pbkdf2_hmac_sha512);
-#define SALT_ALIGN               8
 #define PLAINTEXT_LENGTH         110
 #define SALT_SIZE                sizeof(salt_t)
+#define SALT_ALIGN               sizeof(salt_t*)
```
I see no pointer inside the struct. I think 8 was actually right, but we can use `sizeof(uint64_t)` for clarity.

Was there any issue that prompted you to make a change here? I wonder if there's something I overlook.

magnumripper replied:
> Was there any issue that prompted you to make a change here? I wonder if there's something I overlook.

No I only saw the warning from fde on a 32-bit build, but looked for others that had a fixed 8. This was the only one and like I said I thought it looked wrong.

> I see no pointer inside the struct. I think 8 was actually right, but we can use `sizeof(uint64_t)` for clarity.

Oh, I'm always confused by these things, that's why I shouldn't try to fix these things. You are obviously right.

magnumripper commented:
Now that I'm set straight I'll try to address the other ones we have and amend their fixes.
#3677, #4532

solardiz reviewed (approved):
The revised changes look correct to me.

magnumripper committed:
Review SALT_ALIGN and BINARY_ALIGN in pbkdf2-hmac-sha*

pbkdf2-hmac-sha1 and pbkdf2-hmac-sha256 were using ARCH_WORD where
32-bit was intended. pbkdf2-hmac-sha512 had a too large SALT_ALIGN and
a too small BINARY_ALIGN.

magnumripper committed:
SSPR formats: Fix alignment warnings and move to common code

Closes #4532

solardiz reviewed (approved):
The changes make sense to me. I wonder how the pbkdf2-hmac-sha512 format worked on archs requiring alignment when its BINARY_ALIGN was only 4, so any copies of its binary wouldn't be 8-byte aligned half of the time. Perhaps this suggests we don't currently rely on this alignment in this format's `cmp_all()` and `cmp_one()`, at least not on those archs? I didn't check. Anyway, let's have the alignment.

This commit message is subtly wrong:

> Anything using ARCH_WORD is suspect, looks like it was often mistaken
for normally being 32-bit (well it was, in the old days)

`ARCH_WORD` never implied exactly 32-bit, not even "in the old days" pre-dating jumbo. The very reason I introduced `ARCH_WORD` was to have a type for the machine's native word size (usually same as `long`, but not necessarily), and JtR already ran on some 64-bit platforms back then (1990s).

I suggest you keep only the "Anything using ARCH_WORD is suspect" portion.

magnumripper commented:
> The changes make sense to me. I wonder how the pbkdf2-hmac-sha512 format worked on archs requiring alignment when its BINARY_ALIGN was only 4, so any copies of its binary wouldn't be 8-byte aligned half of the time. Perhaps this suggests we don't currently rely on this alignment in this format's `cmp_all()` and `cmp_one()`, at least not on those archs? I didn't check. Anyway, let's have the alignment.

IIRC, only the OpenCL format would actually use `uint64_t` in `cmp_*()` while the CPU format uses `memcmp()`. Non-x86 archs running OpenCL is uncommon (until now with the Silicon M1, but even that is mostly capable of unaligned now).

> This commit message is subtly wrong:
> 
> > Anything using ARCH_WORD is suspect, looks like it was often mistaken
> > for normally being 32-bit (well it was, in the old days)
> 
> `ARCH_WORD` never implied exactly 32-bit, not even "in the old days" pre-dating jumbo. The very reason I introduced `ARCH_WORD` was to have a type for the machine's native word size (usually same as `long`, but not necessarily), and JtR already ran on some 64-bit platforms back then (1990s).

What I meant with "normally" is ten years ago 32-bit was still very common (especially for Windows, so for Jim) and `ARCH_WORD` would often end up as 32. Surely it was used by mistake here. I can't see a reason to change the wording.

magnumripper commented:
Rebased to keep the pbkdf2-hmac* stuff in a single commit, so I did change the wording now ;-)
